### DISCUSSION TITLE
泡泡DNS和网关搭配使用时，网关挂逼了DNS是如何应对的？

### DISCUSSION THREAD

**Discussion by simon-yanghl** [kkkgo/PaoPaoDNS#115] [想法]
### 在提交之前，请确认

- [X] 我已经尝试执行test.sh并搜索过Issue和discussions和文档，但没有找到相关问题。
- [X] 我正在使用最新的docker镜像版本（可以尝试`docker pull sliamb/paopaodns:latest`后重新创建容器）。

### 脚本自检日志

```txt
略
```

### 问题描述和复现步骤

在使用泡泡网关搭配DNS的时候遇到一个问题，如果网关挂逼了，不管是死机还是掉订阅还是怎么样，DNS还是解析一个fakeip给网关，这个时候不管是正经外网还是不正经的都一刀切了。请问DNS和网关之间有通信互相查询是否正常么？是否可以增加一个网关故障自动切换DNS功能到普通递归DNS不解析fakeip呢？

**Comment by kkkgo** [comment]
网关故障切换那个是路由层面的事，DNS做不到；
DNS的话是按你的设定的规则走的。你可以在前面再套一层adguradhome或者mosdns或者别的什么DNS来实现故障转移。

**Reply by kkkgo** [reply]
可以参考 https://github.com/kkkgo/PaoPaoGateWay/discussions/72#discussioncomment-8841955

**Comment by kkkgo** [comment]
实际上，给custom_forward加上“故障转移”带来的潜在的问题远比收益多。比如很多浏览器/APP并不遵守DNS的实际ttl，在拿到真实IP之后会缓存很长一段时间，又或者递归失败、IPV6、其他自定义选项冲突等问题。实际上这种需求往往是伪需求，并没有一个明确的目标，如果确定哪些不需要走代理的域名，完全可以添加到nocn_list，这样就不会走代理，想走代理也可以添加对应真实IP段到静态路由，然后做路由策略故障转移。
故障转移更适合在路由上或者网关上做，比如启动两个ppgw，一个是直接free出站，然后在路由器上做故障转移路由策略。或者在自己写clash配置文件的规则实现故障转移到DIRECT。

**Reply by simon-yanghl** [reply]
我还是不太理解
如果套一层adg之类的，是如何实现故障转移的？
在我的理解里，主路由负责上网，泡泡dns负责dns解析，泡泡网关负责翻墙
泡泡网关故障，泡泡dns还是依然解析fakeip出来么？
如果是的话，agd不是依然获得一个fakeip么
那adg如何知道fakeip无法访问（泡泡网关已经挂了），从而返回其他的DNS结果呢？

**Reply by kkkgo** [reply]
设置为并行请求解析，当ppgw挂掉无法给出fakeip的时候，ppdns查询就会超时，此时另一个延迟较大的DNS的结果就会被采纳。

**Reply by kkkgo** [reply]
当然，更优雅的做法是在主路由上做DNS代理，就像我连接里面说的在设定的超时时间内查询不到结果自动转移到其他DNS并降低ttl值。
但无论如何，想要获得“故障转移”都是有副作用的，比如某些域名验证时间太长被其他DNS结果顶掉，又或者手动重启一下ppgw，就会造成前面说的“很多浏览器/APP并不遵守DNS的实际ttl，在拿到真实IP之后会缓存很长一段时间”。

**Reply by simon-yanghl** [reply]
啊？是不是说ppdns给出fakeip的前提是ppgw正常工作？如果ppgw挂了ppdns对外网的解析会超时？
那这里涉及不涉及到ppdns自己的缓存问题？比如我在一秒前最后更新了一次google的ip，然后ppgw挂了，ppdns会因为缓存刚更新而导致没发觉ppgw失效？

我现在打算路由功能硬件实现，dns和翻墙留给软路由。那这个时候主路由需要一定的功能才能实现DNS代理吧？如果在软路由上做adg之类的也是AIB……全炸。

那么话说回来了，您有没有推荐的物美价廉的主路由可以用？需求是双万兆（起码双2.5G，我是双线接入总带宽2.5G+20%浮动）+支持端口复用的多拨（为了配合双线）……

**Reply by kkkgo** [reply]
FAKEIP返回的ttl一般就是3秒或者1秒，ppdns对于custom_forward是尊重ttl的，并不会持久缓存。
其实能进ssh的路由器都可以，反正就是跑二进制。
xdr6088双2.5G刷immortalwrt还可以，联发科旗舰830硬件加速不差，可以参考https://cmi.hanwckf.top/p/immortalwrt-mt798x/ 
ROS当然也可以实现这种功能，但官方的硬路由就不是那么物美价廉。
双10G的qnap的301w，或者配置更高的322都可以。官方openwrt支持。

**Reply by kkkgo** [reply]
从功能定义上来说，custom_forward功能是自定义转发，转发失败自然超时无结果。
但“ppdns给出fakeip的前提是ppgw正常工作”不一定对，能给出fakeip只能说明ppgw并没有死机以及clash核心正常加载配置工作中，至于你的节点能不能访问那是另一回事，这属于DNS本身无法预知的未来。

**Reply by kkkgo** [reply]
我前面说的DNS代理其实是针对docker本身挂掉的场景的，你也可以点进去连接看讨论，实现主路由没挂其他全挂了都可以上网是可以的。

**Reply by simon-yanghl** [reply]
看了一圈确实没太满意的硬件，ROS和UBNT都太贵，TP的商用据说又很脑残……
